tests/equality/equality.cpp

```diff
@@ -0,0 +1,995 @@
+    libcellml::ResetPtr r3 = libcellml::Reset::create();
+    libcellml::ResetPtr r4 = libcellml::Reset::create();
+    libcellml::ResetPtr r5 = libcellml::Reset::create();
+    libcellml::ResetPtr r6 = libcellml::Reset::create();
+
+    r1->setOrder(5);
+    r2->setOrder(3);
+    r3->setOrder(7);
+    r4->setOrder(5);
+    r5->setOrder(3);
+    r6->setOrder(7);
+
+    c1->addReset(r1);
+    c1->addReset(r2);
+    c1->addReset(r3);
+
+    c2->addReset(r4);
+    c2->addReset(r6);
+    c2->addReset(r5);
+
+    EXPECT_TRUE(c1->equal(c2));
+}
+
+TEST(Equality, componentEqualDifferentOrderVariables)
+{
+    libcellml::ComponentPtr c1 = libcellml::Component::create("c");
+    libcellml::ComponentPtr c2 = libcellml::Component::create("c");
+
+    libcellml::VariablePtr v1 = libcellml::Variable::create("vA");
+    libcellml::VariablePtr v2 = libcellml::Variable::create("vB");
+    libcellml::VariablePtr v3 = libcellml::Variable::create("vC");
+    libcellml::VariablePtr v4 = libcellml::Variable::create("vA");
+    libcellml::VariablePtr v5 = libcellml::Variable::create("vB");
+    libcellml::VariablePtr v6 = libcellml::Variable::create("vC");
+
+    c1->addVariable(v1);
+    c1->addVariable(v2);
+    c1->addVariable(v3);
+
+    c2->addVariable(v6);
+    c2->addVariable(v4);
+    c2->addVariable(v5);
+
+    EXPECT_TRUE(c1->equal(c2));
+}
+
+TEST(Equality, componentEqualDifferentOrderComponents)
+{
+    libcellml::ComponentPtr c1 = libcellml::Component::create("c");
+    libcellml::ComponentPtr c2 = libcellml::Component::create("c");
+
+    libcellml::ComponentPtr cChild1 = libcellml::Component::create("child1");
+    libcellml::ComponentPtr cChild2 = libcellml::Component::create("child2");
+    libcellml::ComponentPtr cChild3 = libcellml::Component::create("child3");
+    libcellml::ComponentPtr cChild4 = libcellml::Component::create("child1");
+    libcellml::ComponentPtr cChild5 = libcellml::Component::create("child2");
+    libcellml::ComponentPtr cChild6 = libcellml::Component::create("child3");
+
+    c1->addComponent(cChild1);
+    c1->addComponent(cChild2);
+    c1->addComponent(cChild3);
+
+    c2->addComponent(cChild5);
+    c2->addComponent(cChild6);
+    c2->addComponent(cChild4);
+
+    EXPECT_TRUE(c1->equal(c2));
+}
+
+TEST(Equality, componentEqualWithComponentHierarchy)
+{
+    libcellml::ComponentPtr c1 = libcellml::Component::create("c");
+    libcellml::ComponentPtr c2 = libcellml::Component::create("c");
+
+    libcellml::ComponentPtr cChild1 = libcellml::Component::create("child1");
+    libcellml::ComponentPtr cChild2 = libcellml::Component::create("child1");
+
+    libcellml::ComponentPtr cChild11 = libcellml::Component::create("child11");
+    libcellml::ComponentPtr cChild21 = libcellml::Component::create("child11");
+
+    c1->addComponent(cChild1);
+    c2->addComponent(cChild2);
+
+    cChild1->addComponent(cChild11);
+    cChild2->addComponent(cChild21);
+
+    EXPECT_TRUE(c1->equal(c2));
```

Comment:
Making a comment so I remember to check:  is a child component equal to a non-child component?  ie: in this example, if you had another component that wasn't encapsulated, is it the same as one that is? ie: Does/should position in the hierarchy matter?
```
libcellml::ComponentPtr cFakeChild31 = libcellml::Component::create("child11");
EXPECT_WHAT(cFakeChild->equal(cChild11);
```